designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,452 @@
+{
+  ... Some other info ...
+  =======================
+  "results": [
+    {
+      "level": "error",
+      "message": {
+        "text": "'a' is not defined."
+      },
+      "locations": [
+        {
+          "physicalLocation": {
+            "artifactLocation": {
+              "uri": "file:///C:/Users/yiweiding/Documents/projects/test/1.js",
+              "index": 0
+            },
+            "region": {
+              "startLine": 1,
+              "startColumn": 1
+            }
+          }
+        }
+      ],
+      "ruleId": "no-undef",
+      "ruleIndex": 0
+    }
+  ]
+}
+```
+
+JSON is a more popular format for the users of ESLint:
+
+```
+.\node_modules\.bin\eslint.cmd 1.js -f json -o 1.json
+```
+
+```json
+JSON format
+[
+  {
+    "filePath":"C:\\Users\\yiweiding\\Documents\\projects\\test\\1.js",
+    "messages":[
+      {
+        "ruleId":"no-undef",
+        "severity":2,
+        "message":"'a' is not defined.",
+        "line":1,
+        "column":1,
+        "nodeType":"Identifier",
+        "messageId":"undef",
+        "endLine":1,
+        "endColumn":2
+      }
+    ],
+    "errorCount":0,
+    "warningCount":0,
+    "fixableErrorCount":0,
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+Figure Data Flow below is a high-level understanding of the current design.
+
+![ESLint](ESLint.png)
+
+- When rules are suppressed in-source (`eslint-disable` inside of a file), ESLint will still run the rules but then remove any lint messages that were triggered with the specified rule(s).
+
+### What we need
+
+As mentioned above, what we need is to allow engineers to provide justification in the suppression and output the suppression information.
+
+To minimize changes, we propose to use the `-- Description` part of the suppression as "Justification". More specifically, for directive suppression (or inline directive comments in ESLint), the description after two or more consecutive dashes can be treated as the justification.
+
+For example:
+
+```javascript
+/* eslint-disable no-undef -- Justification example */
+a = 1;
+/* eslint-enable no-undef */
+```
+```javascript
+console.log("foo"); // eslint-disable-line no-console -- Justification example
+```
+```javascript
+// eslint-disable-next-line no-console -- Justification example
+console.log("bar");
+```
+
+Then we may get:
+
+```json
+{"kind": "directive", "justification": "Justification example"}
+```
+
+If a violation is inline-suppressed multiple times, all suppression entries will be recorded, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852). So `unusedDisableDirectives` might always be empty. A suppression list would be preferred to gather all suppression information of a violation:
+
+```json
+suppressions: [
+  {"kind": "directive", "justification": "Justification message 1"},
+  {"kind": "directive", "justification": "Justification message 2"}
+]
+```
```

Comment:
We currently don't have this info - only one directive applies, while the others are marked and reported as unused. I think this is an edge case that doesn't warrant the added complexity and confusion that would be caused by directives that are at the same time used (as their justifications appear here) and unused (as they're reported as such).

Reply:
Yes, currently if a violation is suppressed by multiple directives, the first suppression will work while the rest of them will be treated as `unusedDisableDirectives`, as described in the section _Suppression in ESLint_.

What we propose to do is that when using the new option `--track-suppressions`, these suppressions for one violation could be gathered into a suppression list. Without the option, everything works as it is currently. This changes in code would be in `function applyDirectives(options)`.

Reply:
ESLint already marks redundant directives as "unused", and that's observable when the [`--report-unused-disable-directives`](https://eslint.org/docs/user-guide/command-line-interface#--report-unused-disable-directives) flag is used. Furthermore, they cause exit code 1 if this flag is used.

What should be the behavior when both flags are used at the same time? It doesn't make sense to report a directive as unused and to include the same directive in the list of directives that have applied.

Reply:
If `--track-suppression` used, directives that do suppress violations will be considered as "used". But those don't suppress violations will still be tagged as "unused".

Fox example:

```javascript
/* eslint-disable eqeqeq -- Justification 1 */
// eslint-disable-next-line no-undef -- Justification 2
a=1; // eslint-disable-line no-undef -- Justification 3
```

In this code, `// eslint-disable-next-line no-undef -- Justification 2` and `// eslint-disable-line no-undef -- Justification 3` are used to suppress the violation "'a' is not defined". `/* eslint-disable eqeqeq -- Justification 1 */` does nothing so it will be "unused" and be reported when `--report-unused-disable-directives` is used.

I'll update the RFC to clarify it if you are good to this design.

Reply:
Hi @mdjermanovic , we will keep the current behavior. In this example:

```javascript
// eslint-disable-next-line no-undef -- Justification 1
a=1; // eslint-disable-line no-undef -- Justification 2
```

the unused-error should still be throw out if we run `eslint foo.js --report-unused-disable-directives --track-suppressions`.

And we'd like to gather both used and unused suppressions, i.e., we will get a suppression list with 2 items for the above code if we run `eslint foo.js --track-suppressions`:

```json
"suppressions": [
  { "kind": "directive", "justification": "Justification 1" },
  { "kind": "directive", "justification": "Justification 2" }
]
```

Reply:
> ```js
> // eslint-disable-next-line no-undef -- Justification 1
> a=1; // eslint-disable-line no-undef -- Justification 2
> ```
>
> the unused-error should still be throw out if we run `eslint foo.js --report-unused-disable-directives --track-suppressions`.

What would be in `"suppressions": []` in this case?

Reply:
Same as `eslint foo.js --track-suppressions`.

`--report-unused-disable-directives` will report the "unused" error, `--track-suppressions` will gather and output suppressions. They would be independent, right?

Reply:
> `--report-unused-disable-directives` will report the "unused" error, `--track-suppressions` will gather and output suppressions. They would be independent, right?

Yes, in that case the options would work independently of each other, which looks good.

What still looks confusing is that the same directive appears as both used and unused in the same output.

Reply:
The directive is suppressing but is unused at the same time 😄